### ISSUE TITLE
qx.util.Uri.parseUri() goes wrong when @ in query

### ISSUE THREAD

**Issue by mrfeet** [qooxdoo/qooxdoo#9402]
A query string containing the '@' character is interpreted wrong, and everything after the @ is considered to be the host name.

Running this command:
`qx.util.Uri.parseUri("http://localhost/?separator=@")`
results in
>{anchor: "", query: "", file: "", directory: "", path: "", …}
>anchor:""
>authority:"localhost/?separator=@"
>directory:""
>file:""
>host:""
>password:""
>path:""
>port:""
>protocol:"http"
>query:""
>queryKey:{}
>relative:""
>source:"http://localhost/?separator=@"
>user:"localhost/?separator="
>userInfo:"localhost/?separator="

Where there is a value for 'user' and 'query' is empty.

Changing the '@' into 'a' results in something correct:
>{anchor: "", query: "separator=a", file: "", directory: "/", path: "/", …}
>anchor:""
>authority:"localhost"
>directory:"/"
>file:""
>host:"localhost"
>password:""
>path:"/"
>port:""
>protocol:"http"
>query:"separator=a"
>queryKey:{separator: "a"}
>relative:"/?separator=a"
>source:"http://localhost/?separator=a"
>user:""
>userInfo:""

Where 'user' is empty as expected and 'query' contains the query

**Comment by mrfeet** [comment]
The problem is in the regular expressions used for parsing the url. Both the strict and loose one allow too many characters in front of the '@' that is a separator between userid (and optional password) and the domain/ip address.

**Comment by mrfeet** [comment]
I have supplied a pull request with the proper regex:

https://github.com/qooxdoo/qooxdoo/pull/9403

**Comment by derrell** [comment]
I'm not sure this is quite proper. I believe that using an `@` at that point in a URL is not legal; it should be encoded. My basis for this is in this text of RFC3986:

```
2.2.  Reserved Characters

   URIs include components and subcomponents that are delimited by
   characters in the "reserved" set.  These characters are called
   "reserved" because they may (or may not) be defined as delimiters by
   the generic syntax, by each scheme-specific syntax, or by the
   implementation-specific syntax of a URI's dereferencing algorithm.
   If data for a URI component would conflict with a reserved
   character's purpose as a delimiter, then the conflicting data must be
   percent-encoded before the URI is formed.

      reserved    = gen-delims / sub-delims

      gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"

      sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="

   The purpose of reserved characters is to provide a set of delimiting
   characters that are distinguishable from other data within a URI.
   URIs that differ in the replacement of a reserved character with its
   corresponding percent-encoded octet are not equivalent.  Percent-
   encoding a reserved character, or decoding a percent-encoded octet
   that corresponds to a reserved character, will change how the URI is
   interpreted by most applications.  Thus, characters in the reserved
   set are protected from normalization and are therefore safe to be
   used by scheme-specific and producer-specific algorithms for
   delimiting data subcomponents within a URI.

   A subset of the reserved characters (gen-delims) is used as
   delimiters of the generic URI components described in Section 3.  A
   component's ABNF syntax rule will not use the reserved or gen-delims
   rule names directly; instead, each syntax rule lists the characters
   allowed within that component (i.e., not delimiting it), and any of
   those characters that are also in the reserved set are "reserved" for
   use as subcomponent delimiters within the component.  Only the most
   common subcomponents are defined by this specification; other
   subcomponents may be defined by a URI scheme's specification, or by
   the implementation-specific syntax of a URI's dereferencing
   algorithm, provided that such subcomponents are delimited by
   characters in the reserved set allowed within that component.

   URI producing applications should percent-encode data octets that
   correspond to characters in the reserved set unless these characters
   are specifically allowed by the URI scheme to represent data in that
   component.  If a reserved character is found in a URI component and
   no delimiting role is known for that character, then it must be
   interpreted as representing the data octet corresponding to that
   character's encoding in US-ASCII.
```

The `@` is a generic delimiter. The second-to-last paragraph above tells me that generic delimiters are not to be used except in the exact location for which they are used as the delimiter.

**Comment by derrell** [comment]
That said, I'm not positive that I'm opposed to "fixing" it. A general concept in network programming is "be strict in what you send, but be liberal in what you accept." The latter means that if you can figure out what the sender actually meant, do it rather than barfing.

**Comment by mrfeet** [comment]
Well, a user can type it in the URL, and the browser will simply pass it as is.

**Comment by derrell** [comment]
Right, which is why your "fix", although really technically wrong, may be appropriate to have here. What other types of illegal input could a user provide, though, that we should just accept? I'll bet there are many similar things that just haven't come up yet because no user happens to have entered that incorrect URL. How far do we go to prevent the naive user from shooting himself in the foot due to his own ignorance? :-)

Let's see what others think about this. I suspect it ends up being "a good thing".

**Comment by mrfeet** [comment]
I guess the last paragraph actually states it that one should simply accept it:

> If a reserved character is found in a URI component and
>    no delimiting role is known for that character, then it must be
>    interpreted as representing the data octet corresponding to that
>    character's encoding in US-ASCII.

**Comment by level420** [comment]
Fixed via #9403
